src/crypto/ocb_openssl.cc

```diff
@@ -1,1206 +1,189 @@
-            ae_encrypt(&ctx,NULL,pt,0,pt+first,second,ct,NULL,AE_PENDING);
-            ae_encrypt(&ctx,NULL,pt,0,pt+first+second,third,ct,NULL,AE_FINALIZE);
-            memcpy(next,ct,16);
-            next = next+16;
-        }
-
-    }
-    nonce[11] = 0;
-    ae_encrypt(&ctx,nonce,NULL,0,val_buf,next-val_buf,ct,tag,AE_FINALIZE);
-    pbuf(tag,16,0);
-    if (memcmp(valid,tag,16) == 0)
-    	printf("Vectors: PASS\n");
-    else
-    	printf("Vectors: FAIL\n");
-
-
-    /* Encrypt/Decrypt test */
-    for (i = 0; i < 128; i++) {
-        int first = ((i/3)/(BPI*16))*(BPI*16);
-        int second = first;
-        int third = i - (first + second);
-
-        nonce[11] = i%128;
-
-        if (1) {
-            len = ae_encrypt(&ctx,nonce,val_buf,i,val_buf,i,ct,tag,AE_FINALIZE);
-            len = ae_encrypt(&ctx,nonce,val_buf,i,val_buf,-1,ct,tag,AE_FINALIZE);
-            len = ae_decrypt(&ctx,nonce,ct,len,val_buf,-1,pt,tag,AE_FINALIZE);
-            if (len == -1) { printf("Authentication error: %d\n", i); return; }
-            if (len != i) { printf("Length error: %d\n", i); return; }
-            if (memcmp(val_buf,pt,i)) { printf("Decrypt error: %d\n", i); return; }
-        } else {
-            len = ae_encrypt(&ctx,nonce,val_buf,i,val_buf,i,ct,NULL,AE_FINALIZE);
-            ae_decrypt(&ctx,nonce,ct,first,val_buf,first,pt,NULL,AE_PENDING);
-            ae_decrypt(&ctx,NULL,ct+first,second,val_buf+first,second,pt+first,NULL,AE_PENDING);
-            len = ae_decrypt(&ctx,NULL,ct+first+second,len-(first+second),val_buf+first+second,third,pt+first+second,NULL,AE_FINALIZE);
-            if (len == -1) { printf("Authentication error: %d\n", i); return; }
-            if (memcmp(val_buf,pt,i)) { printf("Decrypt error: %d\n", i); return; }
-        }
-
-    }
-    printf("Decrypt: PASS\n");
+int ae_decrypt(ae_ctx *ctx, const void *nonce_ptr, const void *ct_ptr,
+	       int ct_len, const void *ad_ptr, int ad_len, void *pt_ptr,
+	       const void *tag, int final) {
+  const unsigned char *nonce =
+    reinterpret_cast<const unsigned char *>(nonce_ptr);
+  const unsigned char *ct = reinterpret_cast<const unsigned char *>(ct_ptr);
+  const unsigned char* ad = reinterpret_cast<const unsigned char *>(ad_ptr);
+  unsigned char* pt = reinterpret_cast<unsigned char *>(pt_ptr);
+  if (ct_len < ctx->tag_len) {
+    return AE_INVALID;
+  }
+  // If an external tag is not provided, then the tag is assumed to be
+  // the final bytes of the cipher text. Subtract it off now so the
+  // plaintext does not accidentally try to decrypt the AEAD tag (and
+  // then cause an authentication failure).
+  if (tag == NULL) {
+    ct_len -= ctx->tag_len;
+  }
+  // Like encryption, nonce must always be provided and streaming is not supported.
+  if (final != AE_FINALIZE) {
+    return AE_NOT_SUPPORTED;
+  }
+  if (nonce == NULL) {
+    return AE_NOT_SUPPORTED;
+  }
+  if (EVP_DecryptInit_ex(ctx->dec_ctx, /*type=*/NULL, /*impl=*/NULL,
+			 ctx->key, nonce) != 1) {
```

Comment:
Sorry, missed this the first time around: rather than retain `ctx->key`, since you're making both directions of `EVP_CIPHER_CTX` now, just pass the key into `ae_evp_cipher_init` and into `EVP_CipherInit_ex`. (Still NULL IV.) From there you should be able to remove `ctx->key` and instead pass a NULL key in here and `ae_encrypt`.

That should finally only do the per-key setup work at the start, rather than per operation.

Reply:
Sorry, I misunderstood our discussion the other day and thought you said that OpenSSL would recompute the key schedule no matter what, so I didn't put in this optimization. Done.